File: C/21_Binary_Search/BOJ_12738_가장긴증가하는부분수열3.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
main() {
	int n, top = -1;
	scanf("%d", &n);
	int* list = (int*)malloc(sizeof(int) * (n + 1));
	list[++top] = INT_MIN;
	while (n--) {
		int value;
		scanf("%d", &value);
		if (value > list[top]) {
			list[++top] = value;
		}
		else {
			int left = 1, right = top, idx = 1;
			while (left <= right) {
				int mid = (right + left) / 2;
				if (list[mid] >= value) {
					idx = mid;
					right = mid - 1;
				}
				else {
					left = mid + 1;
				}
			}
			list[idx] = value;
		}
	}
	printf("%d", top);
}

File: JAVA/21_Binary_Search/BOJ_12738_가장긴증가하는부분수열3.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        ArrayList<Integer> list = new ArrayList<>();
        list.add(Integer.MIN_VALUE);
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        while (n-- > 0) {
            int value = Integer.parseInt(st.nextToken());
            if (value > list.get(list.size() - 1)) {
                list.add(value);
            } else {
                int left = 1, right = list.size() - 1, idx = 1;
                while (left <= right) {
                    int mid = (left + right) / 2;
                    if (list.get(mid) >= value) {
                        idx = mid;
                        right = mid - 1;
                    } else {
                        left = mid + 1;
                    }
                }
                list.set(idx, value);
            }
        }
        System.out.println(list.size() - 1);
    }
}